zenoh/src/api/session.rs

```diff
@@ -503,104 +489,42 @@ impl Session {
         runtime: Runtime,
         aggregated_subscribers: Vec<OwnedKeyExpr>,
         aggregated_publishers: Vec<OwnedKeyExpr>,
+        owns_runtime: bool,
     ) -> impl Resolve<Session> {
         ResolveClosure::new(move || {
             let router = runtime.router();
-            let state = Arc::new(RwLock::new(SessionState::new(
+            let state = RwLock::new(SessionState::new(
                 aggregated_subscribers,
                 aggregated_publishers,
-            )));
-            let session = Session {
+            ));
+            let session = Session(Arc::new(SessionInner {
+                weak_counter: Mutex::new(0),
                 runtime: runtime.clone(),
-                state: state.clone(),
+                state,
                 id: SESSION_ID_COUNTER.fetch_add(1, Ordering::SeqCst),
-                close_on_drop: true,
-                owns_runtime: false,
+                owns_runtime,
                 task_controller: TaskController::default(),
-            };
+            }));
 
-            runtime.new_handler(Arc::new(admin::Handler::new(session.clone())));
+            runtime.new_handler(Arc::new(admin::Handler::new(session.downgrade())));
 
-            let primitives = Some(router.new_primitives(Arc::new(session.clone())));
-            zwrite!(state).primitives = primitives;
+            let primitives = Some(router.new_primitives(Arc::new(session.downgrade())));
+            zwrite!(session.0.state).primitives = primitives;
 
-            admin::init(&session);
+            admin::init(session.downgrade());
 
             session
         })
     }
 
-    /// Consumes the given `Session`, returning a thread-safe reference-counting
-    /// pointer to it (`Arc<Session>`). This is equivalent to `Arc::new(session)`.
-    ///
-    /// This is useful to share ownership of the `Session` between several threads
-    /// and tasks. It also allows to create [`Subscriber`](crate::pubsub::Subscriber) and
-    /// [`Queryable`](crate::query::Queryable) with static lifetime that can be moved to several
-    /// threads and tasks
-    ///
-    /// Note: the given zenoh `Session` will be closed when the last reference to
-    /// it is dropped.
-    ///
-    /// # Examples
-    /// ```no_run
-    /// # #[tokio::main]
-    /// # async fn main() {
-    /// use zenoh::prelude::*;
-    ///
-    /// let session = zenoh::open(zenoh::config::peer()).await.unwrap().into_arc();
-    /// let subscriber = session.declare_subscriber("key/expression")
-    ///     .await
-    ///     .unwrap();
-    /// tokio::task::spawn(async move {
-    ///     while let Ok(sample) = subscriber.recv_async().await {
-    ///         println!("Received: {:?}", sample);
-    ///     }
-    /// }).await;
-    /// # }
-    /// ```
-    pub fn into_arc(self) -> Arc<Self> {
```

Comment:
@wyfo I'd be tempted to not remove `into_arc` but to deprecate it saying it's no longer needed. In this way we minimise a bit the quantity of code breaking (at least for betas).

Reply:
I was hesitating too. But for any real exemple where the session is stored in a struct, the user will have to update the type, so removing `into_arc` will follow naturally.
Only one-file example just using inference will benefit keeping this method. That's why I've finally removed it.